libcxx/utils/gdb/libcxx/printers.py

```diff
@@ -231,7 +199,7 @@ def __init__(self, val):
 
     def to_string(self):
         """Build a python string from the data whether stored inline or separately."""
-        value_field = _value_of_pair_first(self.val["__r_"])
+        value_field = self.val["__rep_"]
```

Comment:
For the record, this change means that `utils/gdb/libcxx/printers.py` no longer supports objects that were compiled against older versions of libc++. It is of course your call whether you care about that.

When we change layout of libstdc++ types I generally try to ensure that the printers are able to handle the old and new layouts, so that GDB can still do pretty printing for binaries that weren't all compiled with the same GCC release (or which were compiled with the same old GCC release but are dynamically linked to a newer `libstdc++.so.6` at runtime, and so are using the `printers.py` from the newer libstdc++). That isn't trivial to support though, and needs extra tests like https://gcc.gnu.org/cgit/gcc/tree/libstdc++-v3/testsuite/libstdc++-prettyprinters/compat.cc to be sure it keeps working. I'm not requesting that libc++ does it, just noting the consequence of this change which I didn't see mentioned in the PR.

Reply:
I suppose if you're manually loading the printers and calling the registration function, yes you could do that (and switch back and forth if the application contains both definitions). You would need to have both `printer.py` files on disk, in known locations. If you don't have the LLVM source tree present, I'm not sure you'd have the old `printers.py` to load it.

For libstdc++ the printers are loaded automatically when `libstdc++.so.6` is part of the process, so the user doesn't source them manually or register the printers manually. We want it to Just Work.

Reply:
Actually no, it doesn't seem to work. I think you need the `printers.py` sourced and the `register_libcxx_printer_loader` function called before running the process. If printing the type fails, I don't think you can load a different printer later (at least, it doesn't work for me).

So you need to know which printers you're going to want to use before you start debugging the process and pick the right ones.

It seems more likely that users will get the following error and just shrug and assume it doesn't just work :)

`Python Exception <class 'gdb.error'>: There is no member or method named __size_.`

Reply:
FWIW, LLDB does explicitly check for the old and new formats from within the pretty-printers. E.g.,: https://github.com/llvm/llvm-project/blob/7b6e0d9fc3993f3e3df596fd16d97e2ed2e1d0aa/lldb/examples/synthetic/libcxx.py#L743-L749